File: stack-queues/232.js
class MyQueue {
    constructor() {
        this.input = [];
        this.output = [];
        this.peekEl = -1;
    }

    push(x) {
        if (this.input.length === 0) {
            this.peekEl = x;
        }
        this.input.push(x);
    }

    pop() {
        if (this.output.length === 0) {
            while (this.input.length > 0) {
                this.output.push(this.input.pop());
            }
        }
        return this.output.pop();
    }

    peek() {
        if (this.output.length === 0) {
            return this.peekEl;
        }
        return this.output[this.output.length - 1];
    }

    empty() {
        return this.input.length === 0 && this.output.length === 0;
    }
}

File: stack-queues/232.py
# Amortized->O(1)
class MyQueue:
    def __init__(self):
        self.input = []
        self.output = []
        self.peekEl = -1

    def push(self,x):
        if not self.input:
            self.peekEl = x
        self.input.append(x)

    def pop(self):
        if not self.output:
            while self.input:
                self.output.append(self.input.pop())
        val = self.output.pop()
        return val

    def peek(self):
        if not self.output:
            return self.peekEl
        return self.output[-1]

    def empty(self):
        return len(self.input) == 0 and len(self.output) == 0